pkg/cri/resource-manager/policy/builtin/memtier/node.go

```diff
@@ -451,7 +451,7 @@ func (n *numanode) GetPhysicalNodeIDs() []system.ID {
 func (n *numanode) DiscoverSupply() Supply {
 	log.Debug("discovering CPU available at node %s...", n.Name())
 
-	noderes := n.sysnode.CPUSet().Intersection(n.policy.allowed)
+	noderes := n.sysnode.CPUSet().Intersection(n.policy.allowed).Difference(n.policy.reserved)
```

Comment:
I don't think we can do it like this. With this implementation, `ReservedResources[CPU]` would be just another way of expressing `AllowedResources[CPU]`, by using negation. However, the semantics of `ReservedResources` should be 'resources reserved for consumption by containers in the `kube-system` namespace.`. So we should allocate `kube-system` containers to those as opposed to the current approach of just pushing them to the root pool.

Reply:
Ah, sure! (Should have read the docs!) I'll take this back and see what I could do the `kube-system` allocation...

Reply:
> @klihub, I ended up rescoping this PR back to the roots - and beyond. That is, it will test only AvailableResources, and it makes a check on shared pool in the case where allocations look sane.
> 
> With `sed -i s/CONTCOUNT=2 CPU=980m/CONTCOUNT=1 CPU=2500m/' code.var.sh`, however, I'm getting (in interactive mode):
> 
> ```
> run.sh> pp 'cpus["pod1c0"]'                                                        |
> {'cpu10'}
> ```
> 
> yet allocations in cache says:
> 
> ```
> run.sh> pp 'allocations["pod1c0"]'
> {'ColdStart': 0,
>  'Container': '076eac8c-b65f-40da-8f74-c68c7c53e228:pod1c0',
>  'Exclusive': '8-9',
>  'MemType': 'DRAM,PMEM,HBM',
>  'MemoryLimit': {'0': 0, '1': 100000000, '2': 0, '4': 0, '7': 0},
>  'MemoryPool': 'numa node #2',
>  'Memset': '2',
>  'Part': 500,
>  'Pool': 'numa node #2'}
> ```
> 
> That is, it looks like real cpus_allowed do not include exclusive cpus. This happens on the (default) distro `ubuntu-20.04`, so it's probably not broken `runc` this time.

Hmm, I need to reread the code and see if my earlier comment about fractional Guaranteed QoS allocations only in the 1 < cpu req. < 2 range is true. Because it looks to me as if here we'd try to do the same (IOW, allocate 2 CPUs exclusively, then assign the container to run on those + the shared pool with the remaining fractional (.5) CFS share ('Part')...

So what do you have as that container's cpuset.cpus entry ? We don't record the exact shared shared cpuset of a pool per container, only a reference to it via the 'Pool' here together with how big slice of the shared part of that to assign to the container. When assigning CPUs, what should happen is that if the shared part is not 0, then the cpuset should be union(exclusive, pool shared cpuset)...

Reply:
I did a quick test with something like this:
```
...
      resources:
        requests:
          cpu: 2500m
          memory: 200M
        limits:
          cpu: 2500m
          memory: 200M
...
   resources:
        requests:
          cpu: 1500m
          memory: 200M
        limits:
          cpu: 1500m
          memory: 200M
```

That ends up with allocations in the cache like this:
```
  "ce782bf5-6162-4e6d-bc83-fcdf8cb3d745:c0": {
    "Exclusive": "0,12",
    "Part": 500,
    "Container": "ce782bf5-6162-4e6d-bc83-fcdf8cb3d745:c0",
    "Pool": "numa node #0",
    "MemoryPool": "numa node #0",
    "MemType": "DRAM,PMEM,HBM",
    "Memset": "0,4",
    "MemoryLimit": {
      "0": 0,
      "1": 0,
      "2": 200000000,
      "4": 0,
      "7": 0
    },
    "ColdStart": 0
  },
...
  "ce782bf5-6162-4e6d-bc83-fcdf8cb3d745:c1": {
    "Exclusive": "2",
    "Part": 500,
    "Container": "ce782bf5-6162-4e6d-bc83-fcdf8cb3d745:c1",
    "Pool": "numa node #2",
    "MemoryPool": "numa node #2",
    "MemType": "DRAM,PMEM,HBM",
    "Memset": "2,5",
    "MemoryLimit": {
      "0": 0,
      "1": 0,
      "2": 200000000,
      "4": 0,
      "7": 0
    },
    "ColdStart": 0
  },
```

With the NUMA node pools having this state  
```
D: [     policy     ] <post-alloc>        <NUMA node #0>
D: [     policy     ] <post-alloc>          - <numa node #0 capacity: CPU: isolated:60, sharable:0-48:12,72-108:12 (9000m), MemLimit: dram:33177960448, pmem:133143986176>
D: [     policy     ] <post-alloc>          - <numa node #0 allocatable: CPU: isolated:60, sharable:24-48:12,72-108:12 (granted:500m/500m local/subtree, allocatable:6500m), MemLimit: dram:33177960448, pmem:132943986176>
D: [     policy     ] <post-alloc>          - normal memory: 0
D: [     policy     ] <post-alloc>          - HBM memory: 
D: [     policy     ] <post-alloc>          - PMEM memory: 4
D: [     policy     ] <post-alloc>            + <CPU grant for fractional-busybox:c0 from numa node #0: exclusive: 0,12, shared: 24,36,48,72,84,96,108 (500m), MemLimit: pmem:200000000>
...
D: [     policy     ] <post-alloc>        <NUMA node #2>
D: [     policy     ] <post-alloc>          - <numa node #2 capacity: CPU: isolated:6, sharable:18-102:12 (8000m), MemLimit: dram:33816096768, pmem:133143986176>
D: [     policy     ] <post-alloc>          - <numa node #2 allocatable: CPU: isolated:6, sharable:18-102:12 (granted:500m/500m local/subtree, allocatable:7500m), MemLimit: dram:33816096768, pmem:133143986176>
D: [     policy     ] <post-alloc>          - normal memory: 2
D: [     policy     ] <post-alloc>          - HBM memory: 
D: [     policy     ] <post-alloc>          - PMEM memory: 5
D: [     policy     ] <post-alloc>            + <CPU grant for fractional-busybox:c1 from numa node #2: exclusive: 2, shared: 18,30,42,54,66,78,90,102 (500m), MemLimit: pmem:200000000>
D: [     policy     ] <post-alloc>          - parent: <socket #0>
D: [     policy     ] <post-alloc>          - children:

```

The containers end up having these cpusets:
```
r640-1:~ # kubectl exec -ti fractional-busybox -c c0 -- /bin/cat /proc/self/status | grep _allowed_list:
Cpus_allowed_list:      24,36,48,72,84,96,108
Mems_allowed_list:      0,4
r640-1:~ # kubectl exec -ti fractional-busybox -c c1 -- /bin/cat /proc/self/status | grep _allowed_list:
Cpus_allowed_list:      18,30,42,54,66,78,90,102
Mems_allowed_list:      2,5
```

So I think it works as I suspected. I think I'll change the defaults to opt-in from opt-out and file a PR...